buffer/src/main/java/io/netty/buffer/UnpooledHeapByteBuf.java

```diff
@@ -120,29 +120,28 @@ public int capacity() {
     @Override
     public ByteBuf capacity(int newCapacity) {
         checkNewCapacity(newCapacity);
-
-        int oldCapacity = array.length;
         byte[] oldArray = array;
+        int oldCapacity = oldArray.length;
+        if (newCapacity == oldCapacity) {
+            return this;
+        }
+
+        int bytesToCopy;
         if (newCapacity > oldCapacity) {
-            byte[] newArray = allocateArray(newCapacity);
-            System.arraycopy(oldArray, 0, newArray, 0, oldArray.length);
-            setArray(newArray);
-            freeArray(oldArray);
-        } else if (newCapacity < oldCapacity) {
-            byte[] newArray = allocateArray(newCapacity);
-            int readerIndex = readerIndex();
-            if (readerIndex < newCapacity) {
-                int writerIndex = writerIndex();
-                if (writerIndex > newCapacity) {
-                    writerIndex(writerIndex = newCapacity);
-                }
-                System.arraycopy(oldArray, readerIndex, newArray, readerIndex, writerIndex - readerIndex);
-            } else {
+            bytesToCopy = oldCapacity;
+        } else {
+            bytesToCopy = newCapacity;
+            if (readerIndex() > newCapacity) {
                 setIndex(newCapacity, newCapacity);
+            } else if (writerIndex() > newCapacity) {
+                writerIndex(newCapacity);
             }
-            setArray(newArray);
-            freeArray(oldArray);
         }
+        byte[] newArray = allocateArray(newCapacity);
+        //TODO Arrays.copyOf would be preferable here
```

Comment:
Then why not use it ?

Reply:
Was going to write the same...

Reply:
Wouldn't it break subclasses that rely on the use of `allocateArray()` e.g. to [track buffer heap usage](https://github.com/netty/netty/blob/4.1/buffer/src/main/java/io/netty/buffer/UnpooledByteBufAllocator.java#L157) or impl [custom pooling](https://github.com/netty/netty/pull/8015)?

Though I noticed that the `copy()` method currently violates this, I can open another PR to fix that.

Reply:
See #9440 for the other mentioned fix

Reply:
I see.... Makes sense. Can you remove TODO: then ?

Reply:
@normanmaurer done